crates/rustc_codegen_spirv/src/builder/builder_methods.rs

```diff
@@ -1357,132 +1533,179 @@ impl<'a, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'tcx> {
-        let ty = lhs.ty;
-        match self.lookup_type(ty) {
-            SpirvType::Integer(_, _) => {
-                self.emit()
-                    .bitwise_or(ty, None, lhs.def(self), rhs.def(self))
-            }
-            SpirvType::Bool => self
-                .emit()
-                .logical_or(ty, None, lhs.def(self), rhs.def(self)),
-            o => self.fatal(format!(
-                "or() not implemented for type {}",
-                o.debug(ty, self)
-            )),
+    simple_op! {fdiv, float: f_div}
+    simple_op! {fdiv_fast, float: f_div} // fast=normal
+    simple_op! {fdiv_algebraic, float: f_div} // algebraic=normal
+    simple_op! {
+        urem,
+        uint: u_mod,
+        fold_const {
+            uint(a, b) => a.wrapping_rem(b);
         }
-        .unwrap()
-        .with_type(ty)
     }
-    fn xor(&mut self, lhs: Self::Value, rhs: Self::Value) -> Self::Value {
-        assert_ty_eq!(self, lhs.ty, rhs.ty);
-        let ty = lhs.ty;
-        match self.lookup_type(ty) {
-            SpirvType::Integer(_, _) => {
-                self.emit()
-                    .bitwise_xor(ty, None, lhs.def(self), rhs.def(self))
-            }
-            SpirvType::Bool => {
-                self.emit()
-                    .logical_not_equal(ty, None, lhs.def(self), rhs.def(self))
-            }
-            o => self.fatal(format!(
-                "xor() not implemented for type {}",
-                o.debug(ty, self)
-            )),
+    simple_op! {
+        srem,
+        sint: s_rem,
+        fold_const {
+            sint(a, b) => a.wrapping_rem(b);
         }
-        .unwrap()
-        .with_type(ty)
     }
-    fn not(&mut self, val: Self::Value) -> Self::Value {
-        match self.lookup_type(val.ty) {
-            SpirvType::Integer(_, _) => self.emit().not(val.ty, None, val.def(self)),
-            SpirvType::Bool => {
-                let true_ = self.constant_bool(self.span(), true);
-                // intel-compute-runtime doesn't like OpLogicalNot
-                self.emit()
-                    .logical_not_equal(val.ty, None, val.def(self), true_.def(self))
-            }
-            o => self.fatal(format!(
-                "not() not implemented for type {}",
-                o.debug(val.ty, self)
-            )),
+    simple_op! {frem, float: f_rem}
+    simple_op! {frem_fast, float: f_rem} // fast=normal
+    simple_op! {frem_algebraic, float: f_rem} // algebraic=normal
+    simple_shift_op! {
+        shl,
+        int: shift_left_logical,
+        fold_const {
+            int(a, b) => a.wrapping_shl(b as u32);
+        }
+    }
+    simple_shift_op! {
+        lshr,
+        uint: shift_right_logical,
+        fold_const {
+            uint(a, b) => a.wrapping_shr(b as u32);
+        }
+    }
+    simple_shift_op! {
+        ashr,
+        sint: shift_right_arithmetic,
+        fold_const {
+            sint(a, b) => a.wrapping_shr(b as u32);
+        }
+    }
```

Comment:
It should be possible to show in a test that these misbehave - see https://github.com/Rust-GPU/rust-gpu/pull/317#discussion_r2200873033